src/paragraphcommand.js

```diff
@@ -69,10 +72,28 @@ export default class ParagraphCommand extends Command {
 	 * @private
 	 */
 	_updateValue() {
-		const block = this.editor.document.selection.getSelectedBlocks().next().value;
+		const block = this._getSelectedBlock();
+
+		this.value = !!block && block.is( 'paragraph' );
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	_checkEnabled() {
+		const block = this._getSelectedBlock();
+		const schema = this.editor.document.schema;
 
-		if ( block ) {
-			this.value = block.is( 'paragraph' );
-		}
+		return !!block && schema.check( { name: 'paragraph', inside: block.parent.name } ) && !schema.objects.has( block.name );
+	}
+
+	/**
+	 * Returns currently selected block.
+	 *
+	 * @private
+	 * @returns {module:engine/model/element~Element}
+	 */
+	_getSelectedBlock() {
+		return this.editor.document.selection.getSelectedBlocks().next().value;
```

Comment:
Use `first()` from ckeditor5-utils. Also, I think you can repeat this code twice and avoid a method. Or make it a function. Let's not put too many things to prototypes – that makes classes unnecessarily harder to read. OFC, there's nothing particularly wrong in this, but I would do it differently to avoid the minimal bloat it brought.